### PR TITLE
Create consensus component - Closes #6560

### LINKED ISSUE #6560: Create consensus component using processor code
### Description

Create consensus component where it receives and execute blocks with below interface.

dependencies such as below should be resolved in constructor.
```
  chain: Blockchain,
  p2p: P2P,
  stateMachine: StateMachine,
```

```
interface Consensus {
  init(params: InitParams): Promise<void>;
  start(): Promise<void>;
  stop(): Promise<void>;
  syncing(): bool;
  finalizedHeight(): number;
  // execute(block, peerID): Promise<void>;
}

interface InitParams {
  logger: Logger,
  blockchainDB: DB,
  genesisBlock: genesisBlock,
}
```
It should include all the logic from processor except where modules are handled. (this is handled in state machine).
Also, it should include the logic from `transport` where block is received from the  p2p.  

the  consensus folder structure should be as below.
```
/consensus
  consensus.ts
  /synchronizer
```

### Acceptance Criteria
- Unit tests are added
- Interface are following above
- synchronizer source code and tests are duplicated for now

### Additional Information
Related  Issues : #6554 
- system modules will be added under consensus folder later

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #6560 

### How was it solved?

- Move synchronizer to consensus domain
- Move transport logic related to synchronizer to under consensus
- Create consensus component using processor code and the state machine

### How was it tested?

- Adding unit tests

### Additional information
- This needs real state store implementation and block header implementation to work (#6564 and #6581 )

### CODE REVIEW

File: framework/test/unit/node/consensus/endpoint.spec.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+	describe('handleRPCGetBlocksFromId', () => {
+		const DEFAULT_BLOCKS_FROM_IDS_RATE_LIMIT_FREQUENCY = 100;
+
+		it('should apply penalty if call exceeds rate limit', () => {
+			// Arrange
+			const blockIds = codec.encode(getBlocksFromIdRequestSchema, {
+				blockId: getRandomBytes(32),
+			});
+			[...new Array(DEFAULT_BLOCKS_FROM_IDS_RATE_LIMIT_FREQUENCY + 1)].map(async () =>
```
[the remark this answers is not in the document]
Comment (ishantiw) [reply framework/test/unit/node/consensus/endpoint.spec.ts:96]: maybe consistently follow `ArrangeActAssert` in the test, also at other places?

File: framework/test/unit/node/consensus/endpoint.spec.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+			const id = getRandomBytes(32);
+			const validBlock = {
+				ids: [id.toString('hex')],
+			};
+
+			beforeEach(() => {
+				(chain.dataAccess.getHighestCommonBlockHeader as jest.Mock).mockResolvedValue(validBlock);
+			});
+
+			it('should return the result', async () => {
```
Comment (ishantiw) [inline framework/test/unit/node/consensus/endpoint.spec.ts:189]:
```suggestion
			it('should return highest common block header', async () => {
```

**Review by ishantiw — CHANGES_REQUESTED** [review]

File: framework/src/node/consensus/endpoints.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+interface RateTracker {
+	[key: string]: { [key: string]: number };
+}
+
+const DEFAULT_LAST_BLOCK_RATE_LIMIT_FREQUENCY = 10;
+const DEFAULT_COMMON_BLOCK_RATE_LIMIT_FREQUENCY = 10;
+const DEFAULT_BLOCKS_FROM_IDS_RATE_LIMIT_FREQUENCY = 100;
+
+export class Endpoint {
```
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/node/consensus/endpoints.ts:48]: true. consensus API will be like `reducer` so maybe `P2PEndpoint`?

[the remark this answers is not in the document]
Comment (ishantiw) [reply framework/src/node/consensus/endpoints.ts:48]: `P2PEndpoint ` or `NetworkEndpoint` both sounds better

File: framework/src/node/consensus/synchronizer/block_synchronization_mechanism.ts [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+
+		const newTipHasPreference = forkStatus === ForkStatus.DIFFERENT_CHAIN;
+
+		if (!newTipHasPreference) {
+			this._logger.debug(
+				{
+					currentTip: this._chain.lastBlock.header.id,
+					previousTip: tipBeforeApplying.header.id,
+				},
+				'Previous tip of the chain has preference over current tip. Restoring chain from temp table',
```
Comment (ManuGowda) [inline framework/src/node/consensus/synchronizer/block_synchronization_mechanism.ts:206]:
```suggestion
				'Previous tip of the chain has preference over current tip. Restoring chain from temp table.',
```

File: framework/test/unit/node/consensus/consensus.spec.ts [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+
+		it('should verify block header', () => {
+			expect(chain.verifyBlockHeader).toHaveBeenCalledWith(block, expect.anything());
+			expect(consensus['_bft'].verifyBlockHeader).toHaveBeenCalledWith(
+				block.header,
+				expect.anything(),
+			);
+		});
+
+		it('should veryfy block using state machine', () => {
```
Comment (ManuGowda) [inline framework/test/unit/node/consensus/consensus.spec.ts:483]:
```suggestion
		it('should verify block using state machine', () => {
```

**Review by ManuGowda — COMMENTED** [review]
LGTM, if its not a huge problem can we try to move the syncronizer code preserving the history? useful for the future reference

File: framework/src/node/consensus/synchronizer/fast_chain_switching_mechanism.ts [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+			);
+		}
+
+		this._logger.debug(
+			{
+				peerId,
+				fromBlockId: highestCommonBlock.id,
+				toBlockId: receivedBlock.header.id,
+			},
+			'Requesting blocks within ID range from peer',
```
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/node/consensus/synchronizer/fast_chain_switching_mechanism.ts:149]: Here peerID is already logged in the above object

File: framework/src/node/consensus/constants.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+export const EVENT_FORK_DETECTED = 'EVENT_FORK_DETECTED';
+export const EVENT_BLOCK_BROADCAST = 'EVENT_BLOCK_BROADCAST';
+
+export const NETWORK_EVENT_POST_BLOCK = 'postBlock';
```
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/node/consensus/constants.ts:18]: this is the rpc method name, so we cannot cange or capital would also be strange

**Review by ManuGowda — APPROVED** [review]

**Review by ishantiw — APPROVED** [review]

**Pull request merged**